compaction/task_manager_module.cc

```diff
@@ -258,6 +258,88 @@ future<> run_table_tasks(replica::database& db, std::vector<table_tasks_info> ta
+    tasks::task_manager::task_ptr task;
+    sstring keyspace;
+    std::vector<table_info> table_infos;
+
+    keyspace_tasks_info(tasks::task_manager::task_ptr t, sstring ks_name, std::vector<table_info> t_infos)
+        : task(t)
+        , keyspace(std::move(ks_name))
+        , table_infos(std::move(t_infos))
+    {}
+};
+
+future<> run_on_keyspace(sstring op, replica::database& db, std::string keyspace, std::function<future<> (replica::keyspace&)> func) {
+    std::exception_ptr ex;
+    tasks::tmlogger.debug("Starting {} on keyspace {}", op, keyspace);
+    try {
+        co_await func(db.find_keyspace(keyspace));
+    } catch (const replica::no_such_keyspace& e) {
+        tasks::tmlogger.warn("Skipping {} of keyspace {}: {}", op, keyspace, e.what());
+    } catch (...) {
+        ex = std::current_exception();
+        tasks::tmlogger.error("Failed {} of keyspace {}: {}", op, keyspace, ex);
+    }
+    if (ex) {
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
+}
+
+// Run on all tables, skipping dropped tables
+future<> run_on_existing_keyspacess(sstring op, replica::database& db, std::vector<std::string> keyspaces, std::function<future<> (replica::keyspace&)> func) {
+    for (const auto& ks : keyspaces) {
+        co_await run_on_keyspace(op, db, ks, func);
+    }
+}
+
+future<> run_keyspace_tasks(replica::database& db, std::vector<keyspace_tasks_info> keyspace_tasks, seastar::condition_variable& cv, tasks::task_manager::task_ptr& current_task, bool sort) {
+    std::exception_ptr ex;
+
+    // While compaction is run on one table, the size of tables may significantly change.
+    // Thus, they are sorted before each invidual compaction and the smallest keyspace is chosen.
+    while (!keyspace_tasks.empty()) {
+        try {
+            if (sort) {
+                // Major compact smaller tables first, to increase chances of success if low on space.
+                // Tables will be kept in descending order.
+                std::ranges::sort(keyspace_tasks, std::greater<>(), [&] (const keyspace_tasks_info& kti) {
+                    try {
+                        return std::accumulate(kti.table_infos.begin(), kti.table_infos.end(), int64_t(0), [&] (int64_t sum, const table_info& t) {
+                            try {
+                                sum += db.find_column_family(t.id).get_stats().live_disk_space_used;
+                            } catch (const replica::no_such_column_family&) {
+                                // ignore
+                            }
+                            return sum;
+                        });
+                    } catch (const replica::no_such_keyspace&) {
+                        return int64_t(-1);
+                    }
+                });
+            }
+            // Task responsible for the smallest keyspace.
+            current_task = keyspace_tasks.back().task;
+            keyspace_tasks.pop_back();
+            cv.broadcast();
```

Comment:
Where is this waited on?

Reply:
Good point, we need to pass an optional pointer to the condition variable to major_keyspace_compaction_task_impl, and if present, we'll `co_await wait_for_your_turn(_cv, _current_task, _status.id);` there, similar to `table_major_keyspace_compaction_task_impl` with the table tasks.

Reply:
passing pointers to `cv` and `current_task` to `major_keyspace_compaction_task_impl` in next version.
they are both null when it is invoked directly from the `force_keyspace_compaction` api.